### PR TITLE
#165717628 User should be able to attend event from slack

### PR DESCRIPTION
**Pull request by Alameen688**
#### What Does This PR Do?

- This PR gives the user the ability to accept an event from slack
#### Description Of Task To Be Completed

- [x] The slack message sent to a user when an event is created in their category of interest should include an "Attend" button
- [x] When a user clicks the "Attend" button, they should get a response message indicating whether the action was successful or not
#### Any Background Context You Want To Provide?
 - None
#### How can this be manually tested?
- Enable interactivity in the slack app and add a Request URL to the slack app pointing to the endpoint '/api/v1/slack/actions/'
- You might need to add a secure (https) URL as the Request URL. You can use ngrok server to tunnel to localhost then use the secure https URL if testing via localhost
- Follow an interest, you can manually insert this in the `api_interests` table in the db
- Create an event from the client site in a category whose id is the same as that of the interest you followed
- You should get a notification via slack that a new event has been created
- Click the attend button and you should get another message stating that - "You are now attending the event"
#### What are the relevant pivotal tracker stories?
[#165717628](https://www.pivotaltracker.com/story/show/165717628)
#### Screenshot(s)
Screen Shot 2019-05-09 at 3 57 21 PM

### CODE REVIEW

File: server/api/views.py [hunk 1]
```diff
@@ -242,3 +244,52 @@ def get(self, request, *args, **kwargs):
...
+        payload = json.loads(data['payload'])
+        actions = payload.get('actions')
+
+        attend_action, action = self.check_action_type(actions, 'attend_event')
+        if attend_action:
+            event_id = int(action['value'])
+            user_slack_id = payload.get('user').get('id')
+
+            try:
+                event = Event.objects.get(id=event_id)
```
Comment (olusoladavid) [inline server/api/views.py:263]: The logic here appears to be the same logic used in the `AttendEvent` mutation resolver. This is a form of code duplication that could cause problems in the future if someone had to change the implementation in one place and then forgets/does not know to change it in the other place. Could we somehow abstract this logic into a single function?

File: server/api/views.py [hunk 2]
```diff
@@ -242,3 +244,52 @@ def get(self, request, *args, **kwargs):
...
+                        event=event
+                    )
+                    user_attendance.save()
+
+                    message = generate_simple_message(
+                        'You\'ve been successfully registered as an attendee for the event :tada:')
+
+                    notify_user(message, user_slack_id)
+            except (Event.DoesNotExist, AndelaUserProfile.DoesNotExist):
+                raise Http404
```
Comment (olusoladavid) [inline server/api/views.py]: I don't think raising a 404 exception is what we want to do here. What message does the user on Slack get?

File: server/graphql_schemas/event/schema.py [hunk 3]
```diff
@@ -115,13 +115,13 @@ def notify_event_in_slack(category, input, new_event):
...
             event_url = f"{dotenv.get('FRONTEND_BASE_URL')}/{event_id}"
-            message = (f"*A new event has been created in {category.name} "
-                       f"group* \n *Title:* {input.get('title')} \n"
-                       f"*Description:* {input.get('description')} \n "
-                       f"*Venue:* {input.get('venue')} \n"
-                       f"*Date:* {input.get('start_date').date()} \n"
-                       f"*Time:* {input.get('start_date').time()}")
-            blocks = new_event_message(message, event_url)
+            message = (f"*A new event has been created in `{category.name}` group* \n"
+                       f"> *Title:*\t\t\t\t{input.get('title')} \n"
```
Comment (olusoladavid) [inline server/graphql_schemas/event/schema.py]: Nice initiative here. I think we should remove the tabs (I noticed the display is inconsistent and there's too much white space) and leave the block quotes.

File: server/graphql_schemas/event/schema.py [hunk 4]
```diff
@@ -130,7 +130,7 @@ def notify_event_in_slack(category, input, new_event):
                 all_users_attendance.append(new_attendance)
                 if instance.follower.slack_id:
                     slack_response = notify_user(
-                        blocks, instance.follower.slack_id)
+                        blocks, instance.follower.slack_id, text="New upcoming event from Andela socials")
```
Comment (olusoladavid) [inline server/graphql_schemas/event/schema.py:133]: Do we need the `text` arg here?

File: server/api/views.py [hunk 5]
```diff
@@ -242,3 +244,52 @@ def get(self, request, *args, **kwargs):
...
+
+                    user_attendance = Attend(
+                        user=user,
+                        status='attending',
+                        event=event
+                    )
+                    user_attendance.save()
+
+                    message = generate_simple_message(
+                        'You\'ve been successfully registered as an attendee for the event :tada:')
```
Comment (olusoladavid) [inline server/api/views.py]: I am thinking we could have a shorter block-quoted message with an active tone here: `> You\'ve successfully registered for the event :tada:`

File: server/api/views.py [hunk 6]
```diff
@@ -242,3 +244,52 @@ def get(self, request, *args, **kwargs):
...
+                raise Http404
+            except IntegrityError:
+                message = generate_simple_message('You are already an attendee for the event :see_no_evil:')
+                notify_user(message, user_slack_id)
+
+        return Response()
+
+    @staticmethod
+    def check_action_type(actions, action_type):
+        for action in actions:
```
Comment (olusoladavid) [inline server/api/views.py:289]: Do we need this loop if we are returning after the first iteration of the loop?

**Review by olusoladavid — COMMENTED** [review]
Excellent job on this implementation @Alameen688. I made some comments on possible ways we can improve it. Please check them out. Cheers 🥂

File: server/api/views.py (hunk 1 above)
Comment (Alameen688) [reply server/api/views.py:263]: Yes, that is a possible and good approach. I'll do that.

File: server/api/views.py (hunk 2 above)
Comment (Alameen688) [reply server/api/views.py]: I'm supposed to send a user message here. Thanks for pointing that out.

File: server/graphql_schemas/event/schema.py (hunk 3 above)
Comment (Alameen688) [reply server/graphql_schemas/event/schema.py]: Alright then. I think that's okay

File: server/graphql_schemas/event/schema.py (hunk 4 above)
Comment (Alameen688) [reply server/graphql_schemas/event/schema.py:133]: I added it when I noticed the slack notification message always displays - "This content cannot be displayed". Adding this provides a message for the slack popup notification

File: server/api/views.py (hunk 5 above)
Comment (Alameen688) [reply server/api/views.py]: I think I like this also

**Review by olusoladavid — APPROVED** [review]
Great job on this implementation @Alameen688. Works as required.

**Pull request merged**